packages/styled/src/base.js

```diff
@@ -132,27 +132,31 @@ let createStyled: CreateStyled = (tag: any, options?: StyledOptions) => {
         newProps.ref = ref
 
         const ele = React.createElement(finalTag, newProps)
+        let possiblyStyleElement = <></>
         if (!isBrowser && rules !== undefined) {
           let serializedNames = serialized.name
           let next = serialized.next
           while (next !== undefined) {
             serializedNames += ' ' + next.name
             next = next.next
           }
-          return (
-            <>
-              <style
-                {...{
-                  [`data-emotion`]: `${cache.key} ${serializedNames}`,
-                  dangerouslySetInnerHTML: { __html: rules },
-                  nonce: cache.sheet.nonce
-                }}
-              />
-              {ele}
-            </>
+          possiblyStyleElement = (
+            <style
+              {...{
+                [`data-emotion`]: `${cache.key} ${serializedNames}`,
+                dangerouslySetInnerHTML: { __html: rules },
+                nonce: cache.sheet.nonce
+              }}
+            />
           )
         }
-        return ele
+        // Need to return the same number of siblings or else `React.useId` will cause hydration mismatches.
+        return (
+          <>
+            {possiblyStyleElement}
+            {ele}
+          </>
```

Comment:
@mitchellhamilton should we even render those style elements for compat caches? I mean - right now we need to because `extractCritical` "analyzes" the rendered HTML. Is this mainly to handle the default cache of the vanilla Emotion?

With `@emotion/react` (and `@emotion/styled`) each render gets its unique cache by default (well, each top emotion-aware component gets one by default but let's skip over that detail for now). I wonder if in this context we shouldn't skip rendering those style elements altogether (presumably we'll only be able to do that with the new SSR APIs that we've been talking about)? Basically, when using any of our SSR APIs we can expect a custom cache to be used (when using `@emotion/react`) and we should be able to get access to that cache and just grab its content (instead of retrieving that from the rendered HTML). Or am I missing something?

I think the goal with the new SSR APIs is to rely roughly on a similar approach as outlined above, right? We can't fully backport this to the old APIs (they will simply get dropped in the next major version anyway) but I wonder if there is any reason why this has not been done closer to this approach from the start.

Reply:
> should we even render those style elements for compat caches?

I'm pretty sure we don't right now. We retrieve the content from the cache, not the html (we look at the html to see _what_ is used because it was originally made from the case where you're just using a single cache for multiple renders, we don't look at the html for the content of the styles though)

Reply:
Yep - this is kinda what I've meant. That with the new API we wouldn't have to look at the rendered HTML because we could rely on caches being exclusive to renders.